nsls2forge_utils/git_utils.py

```diff
@@ -0,0 +1,163 @@
+    pull_request: bool, optional
+        If true issue pull request, defaults to true
+    fork: bool
+        If true create a fork, defaults to true
+    gh: github3.GitHub instance, optional
+        Object for communicating with GitHub, if None build from $USERNAME
+        and $PASSWORD, defaults to None
+
+    Returns
+    -------
+    recipe_dir : str
+        The recipe directory
+    """
+    gh = ctx.gh
+    # first, let's grab the feedstock locally
+    upstream = feedstock_url(fctx=fctx, protocol=protocol)
+    origin = fork_url(upstream, ctx.github_username)
+    feedstock_reponame = feedstock_repo(fctx=fctx)
+
+    if pull_request or fork:
+        repo = gh.repository(organization, feedstock_reponame)
+        if repo is None:
+            fctx.attrs["bad"] = f"{fctx.package_name}: does not match feedstock name\n"
+            return False
+
+    # Check if fork exists
+    if fork:
+        try:
+            fork_repo = gh.repository(ctx.github_username, feedstock_reponame)
+        except github3.GitHubError:
+            fork_repo = None
+        if fork_repo is None or (hasattr(fork_repo, "is_null") and fork_repo.is_null()):
+            print("Fork doesn't exist creating feedstock fork...")
+            repo.create_fork()
+            # Sleep to make sure the fork is created before we go after it
+            time.sleep(5)
+
+    feedstock_dir = os.path.join(ctx.rever_dir, fctx.package_name + "-feedstock")
+    from conda_forge_tick.git_xonsh_utils import fetch_repo
+
+    if fetch_repo(
+        feedstock_dir=feedstock_dir, origin=origin, upstream=upstream, branch=branch,
+    ):
+        return feedstock_dir, repo
+    else:
+        return None
+
+
+def push_repo(session_ctx, fctx, feedstock_dir, body, repo, title, head, branch,
+              organization='nsls-ii-forge'):
+    """
+    Push a repo up to github
+
+    Parameters
+    ----------
+    feedstock_dir : str
+        The feedstock directory
+    body : str
+        The PR body
+
+    Returns
+    -------
+    pr_json: dict
+        The dict representing the PR, can be used with `from_json`
+        to create a PR instance.
+    """
+    with indir(feedstock_dir), env.swap(RAISE_SUBPROC_ERROR=False):
+        # Setup push from doctr
+        # Copyright (c) 2016 Aaron Meurer, Gil Forsyth
+        token = session_ctx.github_password
+        deploy_repo = (
+            session_ctx.github_username + "/" + fctx.feedstock_name + "-feedstock"
+        )
+        if session_ctx.dry_run:
+            repo_url = f"https://github.com/{deploy_repo}.git"
+            print(f"dry run: adding remote and pushing up branch for {repo_url}")
+        else:
+            doctr_run(
+                [
+                    "git",
+                    "remote",
+                    "add",
+                    f"{organization}_remote",
+                    f"https://{token}@github.com/{deploy_repo}.git",
+                ],
+                token=token.encode("utf-8"),
```

Comment:
Do we need to provide token twice - here and in the URL?

Reply:
Yes, I believe to hide the token from the output as shown here https://github.com/drdoctr/doctr/blob/master/doctr/travis.py#L97

Reply:
Good find :)